components/places/src/bookmark_sync/store.rs

```diff
@@ -728,6 +830,79 @@ impl<'a> Merger<'a> {
         item.validity = SyncedBookmarkValidity::from_u8(row.get("validity")?)?.into();
         Ok(item)
     }
+
+    /// Prunes old merge actions from the `moz_bookmarks_synced_actions` table.
+    fn maybe_prune_old_actions(&self) -> Result<()> {
+        let now = Timestamp::now();
+        let elapsed = get_meta::<Timestamp>(self.store.db, ACTIONS_LAST_PRUNED_AT_META_KEY)?
+            .and_then(|t| now.duration_since(t))
+            .unwrap_or_default();
+        if elapsed >= self.prune_old_actions_after {
+            let tx = if !self.external_transaction {
+                Some(self.store.db.begin_transaction()?)
+            } else {
+                None
+            };
+            let prune_until = now
+                .checked_sub(self.prune_old_actions_after)
+                .unwrap_or_default();
+            self.store.db.execute_batch(&format!(
+                "DELETE FROM moz_bookmarks_synced_actions
+                 WHERE at <= {}",
+                prune_until
+            ))?;
+            put_meta(self.store.db, ACTIONS_LAST_PRUNED_AT_META_KEY, &now)?;
+            if let Some(tx) = tx {
+                tx.commit()?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Fetches a list of items that have been reuploaded with new structure for
+    /// the last 5 syncs. This is captured in logs and telemetry, and suggests
+    /// that we might be in a sync loop with another client (bug 1530145).
+    fn fetch_consecutive_reuploads(&self) -> Result<Vec<ConsecutiveReupload>> {
```

Comment:
I'm going to punt on this for later. There's probably a better way to collect this data, anyway—for example, tracking the number of bookmarks with `BookmarkValidityState::Reupload` and `BookmarkValidityState::Replace` in validation telemetry, along with the other problems.

By themselves, consecutive uploads for the same record aren't bad—they're a symptom of other problems that we can already track—and validation and structure change data seems sufficient for now. Also, this query is pretty big, complicated, scans the entire `moz_bookmarks` table and the `WITH` views for each sync, and doesn't capture everything...an invalid remote item that's unchanged locally will have a `Local` merge state, not `RemoteWithNewStructure`.